app/registries/foreman/setting_manager.rb

```diff
@@ -0,0 +1,67 @@
+module Foreman
+  class SettingManager
+    class << self
+      # Holds all the setting definitions for application
+      def settings
+        @settings ||= {}
+      end
+
+      # Holds all setting categories with their labels
+      def categories
+        @categories ||= { 'general' => N_('General') }
+      end
+
+      def define(context_name, &block)
+        new(context_name).instance_eval(&block)
+      end
+    end
+
+    def initialize(context_name)
+      @context_name = context_name
+    end
+
+    def category(category_name, category_label = nil, &block)
+      self.class.categories[category_name.to_s] ||= category_label
+      CategoryMapper.new(self.class.settings, @context_name, category_name.to_s).instance_eval(&block)
+    end
+
+    class CategoryMapper
+      attr_reader :context_name, :category_name
+
+      def initialize(storage, context_name, category_name)
+        @storage = storage
+        @context_name = context_name
+        @category_name = category_name
+      end
```

Comment:
Passing hash around is not sign of good encapsulation. This should have been a single class.

Reply:
I agree that passing the hash isn't great. I think passing a reference to `self` and using an accessor is cleaner

Reply:
The class is only internal class and my main reasoning behind it is that it won't allow you to define `setting` out of category scope. Without it, it would be bit less readable IMHO.
Though I can just have accessor instead of passing the hash around.

Reply:
My point was that CategoryMapper obviously wants data encapsulated by other class, SettingManager, and I think this smells. I was wondering if this should be a single class. However I missed that this is actually an internal class, that is fine. It is okay.